File: app/src/main/java/org/linuxswords/games/tiltmate/time/TimeFormatter.java
package org.linuxswords.games.tiltmate.time;

import java.util.concurrent.TimeUnit;

public class TimeFormatter
{
    private static final String DISPLAY_TIME_FORMAT = "%d:%02d";
    private static final String LOW_TIME_FORMAT = "%d.%d";
    private static final long LOW_TIME_THRESHOLD_MS = 10000;

    public static String convertMillisIntoDisplayableTime(Long millisUntilFinished)
    {
        if (millisUntilFinished < LOW_TIME_THRESHOLD_MS) {
            long seconds = millisUntilFinished / 1000;
            long tenths = (millisUntilFinished % 1000) / 100;
            return String.format(LOW_TIME_FORMAT, seconds, tenths);
        }

        long minutes = TimeUnit.MILLISECONDS.toMinutes(millisUntilFinished);
        long seconds = TimeUnit.MILLISECONDS.toSeconds(millisUntilFinished);
        return String.format(DISPLAY_TIME_FORMAT,
            minutes,
            seconds % 60
        );
    }

}

File: src/game/time-formatter.ts
const LOW_TIME_THRESHOLD_MS = 10_000;

export function formatTime(ms: number): string {
  if (ms <= 0) return '0.0';

  if (ms < LOW_TIME_THRESHOLD_MS) {
    const seconds = Math.floor(ms / 1000);
    const tenths = Math.floor((ms % 1000) / 100);
    return `${seconds}.${tenths}`;
  }

  const minutes = Math.floor(ms / 60_000);
  const seconds = Math.floor((ms % 60_000) / 1000);
  return `${minutes}:${seconds.toString().padStart(2, '0')}`;
}
